pytket_qir/converter.py

```diff
@@ -0,0 +1,368 @@
+            arguments = [qubit.index[0] for qubit in circuit.qubits] + [
+                bit.index[0] for bit in circuit.bits
+            ]  # Order matters.
+            main_circuit.add_circbox(
+                circ_box, arguments, condition=if_bit(self.conditions[block_name])
+            )
+        return main_circuit
+
+    def apply_contraction(self, block: QirBlock) -> Block:
+        """Attempt to contract blocks recursively."""
+        term = block.terminator
+        has_jump = isinstance(term, QirBrTerminator)
+        curr_block = self.cfg[block.name]
+        curr_block.visited = True
+        if has_jump:
+            next_block = cast(
+                QirBlock, self.module.functions[0].get_block_by_name(term.dest)
+            )
+            # A precondition to apply the rewrite rule.
+            continue_next = len(self.cfg[next_block.name].preds) == 1
+
+            if continue_next:  # Recurse through the chain of jumps.
+                next_block_inst = self.apply_contraction(next_block)
+                return Block(
+                    name=curr_block.name,
+                    succs=next_block_inst.succs,
+                    preds=curr_block.preds,
+                    composition=[block.name] + next_block_inst.composition,
+                )
+            else:  # Return the current block.
+                return curr_block
+        return curr_block
+
+    def collapse_blocks(self) -> None:
+        """Detect and collapse chains of linear blocks."""
+        rewritten_cfg = OrderedDict()
+        for block_name, block in self.rewritten_cfg.items():
+            if not block.visited:
+                qir_block = self.module_function.get_block_by_name(block_name)
+                contracted_block = self.apply_contraction(qir_block)
+                if contracted_block is not block:  # Contractions have occured.
+                    rewritten_cfg[block.name] = contracted_block
+                    # Update successors's predecessors to contracted block.
+                    for succ in contracted_block.succs:
+                        preds = [
+                            pred.replace(contracted_block.composition[-1], block_name)
+                            for pred in self.rewritten_cfg[succ].preds
+                        ]
+                        self.rewritten_cfg[succ].preds = preds
+                    # Update the edges datastructure for later use.
+                    self.edges[contracted_block.name] = self.edges[
+                        contracted_block.composition[-1]
+                    ]
+                    for block in contracted_block.composition[1:]:
+                        del self.edges[block]
+                else:
+                    rewritten_cfg[block_name] = block
+        # Reset visited fields to false for later reuse.
+        for block in rewritten_cfg.values():
+            block.visited = False
+        self.rewritten_cfg = rewritten_cfg
+
+
+def circuit_from_qir(
```

Comment:
Here is still some documentation missing. Maybe it would be good to write a few words for each parameter to point out what they are used for?